src/core/env.ml

```diff
@@ -1,51 +1,76 @@
-(** Scoping environment for variables. *)
+(** Scoping environment for variables.
+    Note that since meta-variable types are closed, the contexts passed to
+    {!val:Eval.whnf} and alike are empty. *)
 
+open Extra
 open Terms
 
 (** Type of an environment, used in scoping to associate names to
-   corresponding Bindlib variables and types. Note that it cannot be
-   implemented by a map as the order is important. *)
-type env = (string * (tvar * tbox)) list
+    corresponding Bindlib variables and types. Note that it cannot be
+    implemented by a map as the order is important. The structure is similar
+    to then one of {!type:Terms.ctxt}, a tuple [(x,a,t)] is a variable [x],
+    its type [a] and possibly its definition [t] *)
+type env = (string * (tvar * tbox * tbox option)) list
 
 type t = env
 
-(** [empty] is the empty environemnt. *)
+(** [empty] is the empty environment. *)
 let empty : env = []
 
-(** [add v a env] extends the environment [env] by mapping the string
-   [Bindlib.name_of v] to [(v,a)]. *)
-let add : tvar -> tbox -> env -> env = fun v a env ->
-  (Bindlib.name_of v, (v, a)) :: env
+(** [add v a t env] extends the environment [env] by mapping the string
+    [Bindlib.name_of v] to [(v,a,t)]. *)
+let add : tvar -> tbox -> tbox option -> env -> env = fun v a t env ->
+  (Bindlib.name_of v, (v, a, t)) :: env
 
 (** [find n env] returns the Bindlib variable associated to the variable  name
     [n] in the environment [env]. If none is found, [Not_found] is raised. *)
 let find : string -> env -> tvar = fun n env ->
-  fst (List.assoc n env)
+  let (x,_,_) = List.assoc n env in x
+
+(** [to_prodbox env t] is like [to_prod] below except that it does not
+    unbox the term. *)
+let to_prodbox : env -> tbox -> tbox = fun env t ->
+  let fn t (_,(x,a,u)) =
+    match u with
+    | Some(u) -> _LLet u a (Bindlib.bind_var x t)
+    | None    -> _Prod a (Bindlib.bind_var x t)
+  in
+  List.fold_left fn t env
```

Comment:
I think this function is useless with what I am going to propose for scoping of let-bindings. It should probably be integrated to `to_prod`.